bugtool/cmd/root.go

```diff
@@ -0,0 +1,237 @@
+		"uptime", "dmesg", "bpftool map show", "bpftool prog show",
+		"iptables -S", "ip6tables -S", "iptables -L -v", "ps",
+		"cat /proc/sys/net/core/bpf_jit_enable", "cat /proc/kallsyms",
+		// Versions
+		"docker version",
+		"kubectl version",
+		// Try to find some useful logs
+		"cat /var/log/upstart/docker.log",
+		"cat /var/log/docker.log", "cat /var/log/daemon.log",
+		"journalctl -u cilium*", "cat /var/log/messages", "journalctl -u kubelet",
+		"kubectl -n kube-system get pods",
+		"kubectl get pods,svc --all-namespaces",
+		"kubectl get version",
+	}
+
+	// Append ethtool links
+	if links, err := netlink.LinkList(); err == nil {
+		for _, link := range links {
+			// query current settings
+			sources = append(sources, fmt.Sprintf("ethtool %s", link.Attrs().Name))
+			// query for driver information
+			sources = append(sources, fmt.Sprintf("ethtool -i %s", link.Attrs().Name))
+		}
+	}
+
+	execs := make(chan bool)
+	runInBackground := func(dir, cmd string) {
+		writeCmdToFile(dir, cmd)
+		execs <- true
+	}
+	for _, cmd := range sources {
+		go runInBackground(cmdDir, cmd)
+	}
+	// Wait for the commands to complete
+	for i := 0; i < len(sources); i++ {
+		<-execs
+	}
+}
+
+func copyCiliumCmds(ciliumDir string) {
+	sources := []string{
+		// Most of the output should come via debuginfo but also adding the other ones for skimming purposes
+		"cilium debuginfo", "cilium config", "cilium bpf tunnel list", "cilium bpf lb list",
+		"cilium bpf endpoint list", "cilium bpf ct list global",
+	}
+
+	for _, cmd := range sources {
+		// Add the host flag if set
+		if len(host) > 0 {
+			cmd = fmt.Sprintf("%s -H %s", cmd, host)
+		}
+		writeCmdToFile(ciliumDir, cmd)
+	}
+}
+
+func execCommand(cmd string, args ...string) (string, error) {
+	fmt.Printf("exec: %s %s\n", cmd, args)
+	output, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		return string(output), nil
+	}
+	return string(output), nil
+}
+
+// writeCmdToFile will execute command and write markdown output to a file
+func writeCmdToFile(cmdDir, prompt string) {
+	// Clean up the filename
+	name := strings.Replace(prompt, "/", " ", -1)
+	name = strings.Replace(name, " ", "-", -1)
+	f, _ := os.Create(filepath.Join(cmdDir, name+".md"))
+
+	// Split the command and arguments
+	split := strings.Split(prompt, " ")
+	argc := len(split)
+	var args []string
+	cmd := split[0]
+
+	if argc > 1 {
+		args = split[1:]
+	}
+
+	// Write prompt as header and the output as body, or error but delete empty output.
+	if output, err := execCommand(cmd, args...); err != nil {
+		fmt.Fprintf(f, fmt.Sprintf("Error while running '%s':  %s\n", prompt, err))
+	} else if strings.Contains(output, "```") { // Already contains Markdown, print as is.
+		fmt.Fprint(f, output)
+	} else if len(output) > 0 {
+		fmt.Fprintf(f, fmt.Sprintf("# %s\n\n```\n%s\n```\n", prompt, output))
+	} else { // Delete empty files
+		os.Remove(f.Name())
+		return
+	}
+	f.Close()
```

Comment:
This might have to be defered earlier since we return early just above. Or we can ditch the return to ensure we get here.